File: project_euler_unified/problem_106/solution.py
"""
Problem 106: Special Subset Sums: Meta-testing

For n=12, count how many subset pairs (B, C) with |B|=|C| need to be
tested for equality, given that the ordering rule already handles |B|!=|C|.

Formula: sum over k=2..n/2 of C(n,2k) * [C(2k,k)/2 - Catalan(k)]
"""

from math import comb

def catalan(k):
    """k-th Catalan number: C(2k,k) / (k+1)"""
    return comb(2 * k, k) // (k + 1)

def solve(n=12):
    total = 0
    for k in range(2, n // 2 + 1):
        choose_2k = comb(n, 2 * k)
        half_choose = comb(2 * k, k) // 2
        cat_k = catalan(k)
        contribution = choose_2k * (half_choose - cat_k)
        total += contribution
    return total

answer = solve(12)
print(answer)

File: project_euler_unified/problem_106/solution.cpp
#include <bits/stdc++.h>
using namespace std;

// Compute binomial coefficient C(n, k)
long long comb(int n, int k) {
    if (k < 0 || k > n) return 0;
    if (k == 0 || k == n) return 1;
    if (k > n - k) k = n - k;
    long long result = 1;
    for (int i = 0; i < k; i++) {
        result = result * (n - i) / (i + 1);
    }
    return result;
}

// Catalan number C_k = C(2k,k) / (k+1)
long long catalan(int k) {
    return comb(2 * k, k) / (k + 1);
}

int main() {
    int n = 12;
    long long total = 0;

    for (int k = 2; k <= n / 2; k++) {
        long long choose_2k = comb(n, 2 * k);
        long long half_choose = comb(2 * k, k) / 2;
        long long cat_k = catalan(k);
        total += choose_2k * (half_choose - cat_k);
    }

    cout << total << endl;
    return 0;
}
